### DISCUSSION TITLE
TypeScript glob imports

### DISCUSSION THREAD

**Discussion by codker** [vercel/next.js#20660] [Help] [answered]
Hello everyone and happy new year!

I need help to understand how I can dynamically import a typescript file in a Next API Route.

Let's say that I want to define validation schemas for my models and that I want to use them in React forms and API routes to validate post requests.

I have several files in `components/schemas` that will match this pattern `*.schema.ts`.

When doing a POST request I want to load all the schemas to search for a match (forget how, it's not important). I can't use a dynamic import or require as the files have not been transpiled yet and I don't want to manually maintain an index.ts in `components/schemas` to export every schema.

Anyone has done something similar and can give me some tips?

Thanks!

**Comment by michrome** [comment, marked as answer]
Hey @liberatiluca I think you might be able to use a dynamic import. You have to restrict the **directory with a hard-coded string** to avoid webpack looking at all directories.

Something like this:

```javascript
const { yourExportName } = await import(`../components/schemas/${schemaFileName}`);
```

But you _can't_  do something like this:

```javascript
const { yourExportName } = await import(join(process.cwd(), "components", "schemas", schemaFileName));
```

**Comment by codker** [comment]
> Hey @liberatiluca I think you might be able to use a dynamic import. You have to restrict the **directory with a hard-coded string** to avoid webpack looking at all directories.
> 
> 
> 
> Something like this:
> 
> 
> 
> ```javascript
> 
> const { yourExportName } = await import(`../components/schemas/${schemaFileName}`);
> 
> ```
> 
> 
> 
> But you _can't_  do something like this:
> 
> 
> 
> ```javascript
> 
> const { yourExportName } = await import(join(process.cwd(), "components", "schemas", schemaFileName));
> 
> ```
> 
> 

That put me in the right direction and I found [babel-plugin-import-glob-array](https://github.com/jescalan/babel-plugin-import-glob-array) that transforms a glob import to multiple imports. Supports both default and named exports and works great!

Thanks 🙏
